programs/stake/src/stake_state.rs

```diff
@@ -595,7 +595,7 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
                 let stake = Stake::new(
                     self.lamports()?.saturating_sub(meta.rent_exempt_reserve), // can't stake the rent ;)
                     vote_account.unsigned_key(),
-                    &vote_account.state()?,
+                    &State::<VoteStateVersions>::state(vote_account)?.convert_to_current(),
```

Comment:
@mvines, I dislike this `state()` function because it requires whoever modifies this code to know that vote accounts are special/need special handling, and can't directly be deserialized into a `VoteState` object like its other counterparts(stake, storage, nonce, etc.). Ideally the `Account` type carries the type of the object it's storing so we can avoid this footgun (even in this PR, hopefully I haven't missed any conversions directly to `VoteState` from account.data), but that's a more sweeping change.

Reply:
https://github.com/solana-labs/solana/blob/b18e4057bb23484757ead16b35f04cad2ac53340/sdk/src/account_utils.rs#L38-L40

We could add a trait that requires an implementation of `fn convert_to_current()` then make it a bound on `T` for `State<T>` and `StateMut<T>` hide it in the `fn state()` implementations